File: iosApp/iosApp/service/BiographyService.swift
import Foundation
import Shared

class BiographyService {
    private let backendApi: BackendApi
    private let tokenWrapper: TokenManagerWrapper

    init(tokenWrapper: TokenManagerWrapper, backendApi: BackendApi) {
        self.tokenWrapper = tokenWrapper
        self.backendApi = backendApi
    }

    private func getAuthData() throws -> (token: String, username: String) {
        guard let token = tokenWrapper.getToken(),
              let username = tokenWrapper.getAccount()?.login else {
                tokenWrapper.logout()
                throw BiographyServiceError.authError
        }
        return (token, username)
    }

    private func handleRequest<T>(_ action: () async throws -> T?) async throws -> T {
        do {
            if let result = try await action() {
                return result
            } else {
                throw BiographyServiceError.loadingError
            }
        } catch {
            tokenWrapper.logout()
            throw error
        }
    }

    func getBiography(account: AccountResponse) async throws -> BiographyResponse {
        let (token, username) = try getAuthData()
        do {
            let biography = try await backendApi.getBiography(token: token, username: account.login)
            if (biography == nil) {
                throw BiographyServiceError.notFoundError
            } else {
                return biography!
            }
        } catch is KotlinIllegalStateException {
            tokenWrapper.logout()
            throw BiographyServiceError.authError
        }
    }

    func createBiography(_ request: BiographyRequest) async throws -> BiographyResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveBiography(token: token, biographyRequest: request, biographyId: nil)
        }
    }

    func updateBiography(_ request: BiographyRequest) async throws -> BiographyResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveBiography(token: token, biographyRequest: request, biographyId: request.id!)
        }
    }

    func getLanguagesByBiography(biographyId: Int) async throws -> [LanguageResponse] {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.getLanguagesByBiography(token: token, biographyId: Int32(biographyId))
        }
    }
    
    func createLanguage(request: LanguageRequest) async throws -> LanguageResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveLanguage(token: token, languageRequest: request, languageId: nil)
        }
    }

    func updateLanguage(request: LanguageRequest) async throws -> LanguageResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveLanguage(token: token, languageRequest: request, languageId: request.id)
        }
    }

    func deleteLanguage(languageId: Int) async throws -> Bool {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.deleteLanguage(token: token, languageId: Int32(languageId)).boolValue
        }
    }

    func getEducationsByBiography(biographyId: Int) async throws -> [EducationResponse] {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.getEducationsByBiography(token: token, biographyId: Int32(biographyId))
        }
    }
    
    func createEducation(request: EducationRequest) async throws -> EducationResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveEducation(token: token, educationRequest: request, educationId: nil)
        }
    }

    func updateEducation(request: EducationRequest) async throws -> EducationResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveEducation(token: token, educationRequest: request, educationId: request.id)
        }
    }

    func deleteEducation(educationId: Int) async throws -> Bool {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.deleteEducation(token: token, educationId: Int32(educationId)).boolValue
        }
    }

    func getProjectsByBiography(biographyId: Int) async throws -> [ProjectResponse] {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.getProjectsByBiography(token: token, biographyId: Int32(biographyId))
        }
    }

    func getSkillsByBiography(biographyId: Int) async throws -> [SkillResponse] {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.getSkillsByBiography(token: token, biographyId: Int32(biographyId))
        }
    }
    
    func createSkill(request: SkillRequest) async throws -> SkillResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveSkill(token: token, skillRequest: request, skillId: nil)
        }
    }

    func updateSkill(request: SkillRequest) async throws -> SkillResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveSkill(token: token, skillRequest: request, skillId: request.id)
        }
    }

    func deleteSkill(skillId: Int) async throws -> Bool {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.deleteSkill(token: token, skillId: Int32(skillId)).boolValue
        }
    }
    
    func createProject(request: ProjectRequest) async throws -> ProjectResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveProject(token: token, projectRequest: request, projectId: nil)
        }
    }

    func updateProject(request: ProjectRequest) async throws -> ProjectResponse {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.saveProject(token: token, projectRequest: request, projectId: request.id)
        }
    }

    func deleteProject(projectId: Int) async throws -> Bool {
        let (token, _) = try getAuthData()
        return try await handleRequest {
            try await backendApi.deleteProject(token: token, projectId: Int32(projectId)).boolValue
        }
    }
}

File: composeApp/src/androidMain/kotlin/org/adastra/curriculum/service/BiographyService.kt
package org.adastra.curriculum.service

import org.adastra.curriculum.BuildConfig
import org.adastra.curriculum.auth.LoginViewModel
import org.adastra.curriculum.client.BackendApi
import org.adastra.curriculum.client.data.BiographyRequest
import org.adastra.curriculum.client.data.BiographyResponse
import org.adastra.curriculum.client.data.EducationRequest
import org.adastra.curriculum.client.data.EducationResponse
import org.adastra.curriculum.client.data.LanguageRequest
import org.adastra.curriculum.client.data.LanguageResponse
import org.adastra.curriculum.client.data.ProjectResponse
import org.adastra.curriculum.client.data.SkillResponse
import org.adastra.curriculum.exception.NotFoundException

class BiographyService(loginViewModel: LoginViewModel) {
    private val baseUrl = BuildConfig.BASE_URL
    private val backendApi = BackendApi(baseUrl)
    private val lvm = loginViewModel
    private val token = lvm.getToken()
    private val username = lvm.getAccount()?.login

    suspend fun getBiography(): BiographyResponse {
        var biography: BiographyResponse? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                biography = backendApi.getBiography(token, username)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            } catch (_: Exception) {
                throw Exception("Chyba při načítání dat. Zkuste to prosím později.")
            }
        }

        if (biography == null) {
            throw NotFoundException()
        } else {
            return biography
        }
    }

    suspend fun createBiography(biographyRequest: BiographyRequest): BiographyResponse {
        var biography: BiographyResponse? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                biography = backendApi.saveBiography(token, biographyRequest, null)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (biography == null) {
            throw Exception("Chyba při načítání dat. Zkuste to prosím později.")
        } else {
            return biography
        }
    }

    suspend fun updateBiography(biographyRequest: BiographyRequest): BiographyResponse {
        var biography: BiographyResponse? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                biography = backendApi.saveBiography(token, biographyRequest, biographyRequest.id)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (biography == null) {
            throw Exception("Chyba při načítání dat. Zkuste to prosím později.")
        } else {
            return biography
        }
    }

    suspend fun getLanguagesByBiography(biographyId: Int): List<LanguageResponse> {
        var languages: List<LanguageResponse>? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                languages = backendApi.getLanguagesByBiography(token, biographyId)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (languages == null) {
            throw Exception("Chyba při načítání dat. Zkuste to prosím později.")
        } else {
            return languages
        }
    }

    suspend fun createLanguage(languageRequest: LanguageRequest): LanguageResponse {
        var language: LanguageResponse? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                language = backendApi.saveLanguage(token, languageRequest, null)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (language == null) {
            throw Exception("Chyba při ukládání dat. Zkontrolujte prosím zadané údaje.")
        } else {
            return language
        }
    }

    suspend fun updateLanguage(languageRequest: LanguageRequest): LanguageResponse {
        var language: LanguageResponse? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                language = backendApi.saveLanguage(token, languageRequest, languageRequest.id)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (language == null) {
            throw Exception("Chyba při ukládání dat. Zkontrolujte prosím zadané údaje.")
        } else {
            return language
        }
    }

    suspend fun deleteLanguage(languageId: Int): Boolean {
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                return backendApi.deleteLanguage(token, languageId)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }
    }

    suspend fun getEducationsByBiography(biographyId: Int): List<EducationResponse> {
        var educations: List<EducationResponse>? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                educations = backendApi.getEducationsByBiography(token, biographyId)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (educations == null) {
            throw Exception("Chyba při načítání dat. Zkuste to prosím později.")
        } else {
            return educations
        }
    }

    suspend fun createEducation(educationRequest: EducationRequest): EducationResponse {
        var education: EducationResponse? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                education = backendApi.saveEducation(token, educationRequest, null)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (education == null) {
            throw Exception("Chyba při ukládání dat. Zkontrolujte prosím zadané údaje.")
        } else {
            return education
        }
    }

    suspend fun updateEducation(educationRequest: EducationRequest): EducationResponse {
        var education: EducationResponse? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                education = backendApi.saveEducation(token, educationRequest, educationRequest.id)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (education == null) {
            throw Exception("Chyba při ukládání dat. Zkontrolujte prosím zadané údaje.")
        } else {
            return education
        }
    }

    suspend fun deleteEducation(educationId: Int): Boolean {
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                return backendApi.deleteEducation(token, educationId)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }
    }

    suspend fun getProjectsByBiography(biographyId: Int): List<ProjectResponse> {
        var projects: List<ProjectResponse>? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                projects = backendApi.getProjectsByBiography(token, biographyId)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (projects == null) {
            throw Exception("Chyba při načítání dat. Zkuste to prosím později.")
        } else {
            return projects
        }
    }

    suspend fun getSkillsByBiography(biographyId: Int): List<SkillResponse> {
        var skills: List<SkillResponse>? = null
        if (token == null || username == null) {
            lvm.logout()
            throw Exception("Chyba při načítání dat. Přihlaste se prosím znovu.")
        } else {
            try {
                skills = backendApi.getSkillsByBiography(token, biographyId)
            } catch (e: IllegalStateException) {
                lvm.logout()
                throw Exception(e.message)
            }
        }

        if (skills == null) {
            throw Exception("Chyba při načítání dat. Zkuste to prosím později.")
        } else {
            return skills
        }
    }
}